orchagent/pfcactionhandler.cpp

```diff
@@ -0,0 +1,466 @@
+    resultFvValues.emplace_back(PFC_WD_QUEUE_STATUS, PFC_WD_QUEUE_STATUS_OPERATIONAL);
+
+    countersTable->set(queueIdStr, resultFvValues);
+}
+
+void PfcWdActionHandler::updateWdCounters(const string& queueIdStr, const PfcWdQueueStats& stats)
+{
+    SWSS_LOG_ENTER();
+
+    vector<FieldValueTuple> resultFvValues;
+
+    resultFvValues.emplace_back(PFC_WD_QUEUE_STATS_DEADLOCK_DETECTED, to_string(stats.detectCount));
+    resultFvValues.emplace_back(PFC_WD_QUEUE_STATS_DEADLOCK_RESTORED, to_string(stats.restoreCount));
+    resultFvValues.emplace_back(PFC_WD_QUEUE_STATUS, stats.operational ?
+                                                     PFC_WD_QUEUE_STATUS_OPERATIONAL :
+                                                     PFC_WD_QUEUE_STATUS_STORMED);
+
+    m_countersTable->set(queueIdStr, resultFvValues);
+}
+
+PfcWdLossyHandler::PfcWdLossyHandler(sai_object_id_t port, sai_object_id_t queue,
+        uint8_t queueId, shared_ptr<Table> countersTable):
+    PfcWdActionHandler(port, queue, queueId, countersTable)
+{
+    SWSS_LOG_ENTER();
+
+    sai_attribute_t attr;
+    attr.id = SAI_PORT_ATTR_PRIORITY_FLOW_CONTROL;
+
+    sai_status_t status = sai_port_api->get_port_attribute(port, 1, &attr);
+    if (status != SAI_STATUS_SUCCESS)
+    {
+        SWSS_LOG_ERROR("Failed to get PFC mask on port 0x%lx: %d", port, status);
+    }
+
+    uint8_t pfcMask = attr.value.u8;
+    attr.id = SAI_PORT_ATTR_PRIORITY_FLOW_CONTROL;
+    attr.value.u8 = pfcMask & ~(1 << queueId);
+
+    status = sai_port_api->set_port_attribute(port, &attr);
+    if (status != SAI_STATUS_SUCCESS)
+    {
+        SWSS_LOG_ERROR("Failed to get PFC mask on port 0x%lx: %d", port, status);
+    }
+}
+
+PfcWdLossyHandler::~PfcWdLossyHandler(void)
+{
+    SWSS_LOG_ENTER();
+
+    sai_attribute_t attr;
+    attr.id = SAI_PORT_ATTR_PRIORITY_FLOW_CONTROL;
+
+    sai_status_t status = sai_port_api->get_port_attribute(getPort(), 1, &attr);
+    if (status != SAI_STATUS_SUCCESS)
+    {
+        SWSS_LOG_ERROR("Failed to get PFC mask on port 0x%lx: %d", getPort(), status);
+        return;
+    }
+
+    uint8_t pfcMask = attr.value.u8;
+    attr.id = SAI_PORT_ATTR_PRIORITY_FLOW_CONTROL;
+    attr.value.u8 = pfcMask | (1 << getQueueId());
+
+    status = sai_port_api->set_port_attribute(getPort(), &attr);
+    if (status != SAI_STATUS_SUCCESS)
+    {
+        SWSS_LOG_ERROR("Failed to set PFC mask on port 0x%lx: %d", getPort(), status);
+        return;
+    }
+}
+
+PfcWdZeroBufferHandler::PfcWdZeroBufferHandler(sai_object_id_t port,
+        sai_object_id_t queue, uint8_t queueId, shared_ptr<Table> countersTable):
+    PfcWdLossyHandler(port, queue, queueId, countersTable)
+{
+    SWSS_LOG_ENTER();
+
+    sai_attribute_t attr;
+    attr.id = SAI_QUEUE_ATTR_BUFFER_PROFILE_ID;
+
+    // Get queue's buffer profile ID
+    sai_status_t status = sai_queue_api->get_queue_attribute(queue, 1, &attr);
```

Comment:
Also we could pre-store the queue buffer profile id, instead of querying every time.

Reply:
Ok, will change in P2 along with PG buffer profile.